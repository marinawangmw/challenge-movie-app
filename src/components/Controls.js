import React from 'react';
import { View, Text, StyleSheet, Image, TouchableOpacity } from 'react-native';
import { useTheme } from '@react-navigation/native';
import { spacing, typography } from '@/theme';
import { strings } from '@/localization';

export const Controls = ({ controlDatas, controlStyles }) => {
  const { colors } = useTheme();
  const styles = { ...defaultStyles, ...controlStyles };

  const renderControls = (control, idx) => (
    <View style={styles.controlWrapper} key={idx}>
      <TouchableOpacity onPress={control.handleControlPress}>
        <Image
          source={control.icon}
          style={[styles.icon, control.label === strings.controls.play && styles.playIcon]}
          resizeMode="cover"
          accessibilityIgnoresInvertColors
        />
      </TouchableOpacity>
      <Text style={[{ color: colors.text }, typography.label]}>{control.label}</Text>
    </View>
  );

  return (
    <View style={[styles.container, { color: colors.text }]}>
      {controlDatas.map((control, idx) => renderControls(control, idx))}
    </View>
  );
};

export const defaultStyles = StyleSheet.create({
  icon: {
    width: 28,
    height: 28,
    margin: spacing.xs,
  },
  playIcon: {
    height: 34,
    marginTop: spacing.xs,
    marginBottom: 2,
  },
});
