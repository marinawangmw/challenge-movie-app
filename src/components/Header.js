import React from 'react';
import { View, StyleSheet, Image } from 'react-native';
import { spacing } from '@/theme';

export const Header = () => {
  return (
    <View style={styles.header}>
      <Image
        source={require('@/assets/logo/logo.png')}
        style={styles.headerImg}
        resizeMode="contain"
        accessibilityIgnoresInvertColors
      />
    </View>
  );
};

const styles = StyleSheet.create({
  header: {
    marginTop: spacing.xl,
    alignItems: 'center',
    justifyContent: 'center',
  },
  headerImg: {
    height: 40,
    width: 76,
    marginTop: spacing.s,
  },
});
