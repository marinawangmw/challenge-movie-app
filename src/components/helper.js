import { Linking } from 'react-native';
import { Config } from 'react-native-config';
import Toast from 'react-native-toast-message';
import { getMovieTrailer } from '@/controllers/MovieController';

export const playTrailer = async (id) => {
  try {
    const trailerData = await getMovieTrailer(id);

    if (trailerData.results.length) {
      const url = Config.VIDEO_EXTERNAL_BASE_URL + trailerData.results[0].key;

      const supported = await Linking.canOpenURL(url);

      if (supported) {
        await Linking.openURL(url);
      } else {
        Toast.show({
          type: 'error',
          text1: "Something's wrong",
          text2: `Don't know how to open this URL: ${url}`,
        });
      }
    } else {
      Toast.show({
        type: 'error',
        text1: 'Oops',
        text2: "There's no trailer found for this movie",
      });
    }
  } catch (e) {
    console.log(e);
  }
};
