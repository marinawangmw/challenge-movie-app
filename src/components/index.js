export { Button } from '@/components/Button';
export { ErrorView } from '@/components/ErrorView';
export { TabBarIcon } from '@/components/TabBarIcon';
export { TextField } from '@/components/TextField';
export { Hero } from '@/components/Hero';
export { Header } from '@/components/Header';
export { MovieList } from '@/components/MovieList';
