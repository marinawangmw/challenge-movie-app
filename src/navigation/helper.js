export const screenOptions = (textColor) => ({
  headerShown: true,
  headerBackTitle: '',
  headerTintColor: textColor,
});
