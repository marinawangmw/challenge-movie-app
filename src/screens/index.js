export { Home } from '@/screens/Home/Home';
export { Login } from '@/screens/Login/Login';
export { Profile } from '@/screens/Profile/Profile';
export { Loading } from '@/screens/Loading/Loading';
export { MovieCollection } from '@/screens/MovieCollection/MovieCollection';
export { MovieDetail } from '@/screens/MovieDetail/MovieDetail';
