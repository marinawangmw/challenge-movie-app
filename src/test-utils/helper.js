export const movieList = {
  movieList: [
    {
      adult: false,
      backdrop_path: '/ziTxBaF1AhbUwcv2IBiFJAf54qD.jpg',
      id: 65086,
      overview: 'The story follows a young lawyer, Arthur Kipps.',
      popularity: 16.334,
      poster_path: '/aJbeDPUwDe9MLpseTFrNTBaQPzf.jpg',
      release_date: '2012-02-03',
      title: 'The Woman in Black',
      vote_average: 6.11,
    },
    {
      adult: false,
      id: 102899,
      overview: 'Armed with the astonishing ability to shrink in scale but increase in strength.',
      popularity: 69.704,
      poster_path: '/rS97hUJ1otKTTripGwQ0ujbuIri.jpg',
      release_date: '2015-07-14',
      title: 'Ant-Man',
      vote_average: 7.088,
    },
  ],
  title: 'test',
};

export const navigation = { navigate: jest.fn(), getParam: jest.fn(), setOptions: jest.fn() };
