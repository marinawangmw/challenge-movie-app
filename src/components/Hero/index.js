import React from 'react';
import { View, ImageBackground, StyleSheet } from 'react-native';
import { useSelector, useDispatch } from 'react-redux';
import { useTheme } from '@react-navigation/native';
import { Config } from 'react-native-config';
import LinearGradient from 'react-native-linear-gradient';
import Labels from './Labels';
import { Header, SpecialBanner, Controls } from '@/components';
import { spacing, typography } from '@/theme';
import { getHeroPoster } from '@/selectors/MovieListSelectors';
import { addIcon, playIcon, infoIcon } from '@/assets';
import { strings } from '@/localization';
import { addToMyList } from '@/actions/MovieListActions';
import { NAVIGATION } from '@/constants';

export const Hero = ({ navigation }) => {
  const { colors } = useTheme();
  const heroPoster = useSelector(getHeroPoster);
  const dispatch = useDispatch();

  const handleMyListIconPress = () => {
    dispatch(addToMyList(heroPoster.posterMovie));
  };

  const handleInfoIconPress = () => {
    navigation.navigate(NAVIGATION.movieDetail, {
      movieDetails: heroPoster.posterMovie,
    });
  };

  const controlDatas = [
    { icon: addIcon, label: strings.controls.myList, handleControlPress: handleMyListIconPress },
    { icon: playIcon, label: strings.controls.play },
    { icon: infoIcon, label: strings.controls.info, handleControlPress: handleInfoIconPress },
  ];

  if (heroPoster && heroPoster.posterMovie) {
    return (
      <View style={[styles.container, { backgroundColor: colors.primary }]}>
        <ImageBackground
          source={{ uri: Config.IMAGE_API_BASE_URL + heroPoster.posterMovie.poster_path }}
          resizeMode="cover"
          style={styles.imgBg}
        >
          <LinearGradient
            locations={[0.0001, 1]}
            colors={['rgba(0, 0, 0, 0.0001)', '#000']}
            style={styles.gradient}
          >
            <Header />
            <View style={styles.bottomContainer}>
              <Labels labels={heroPoster.posterGenres} />
              <SpecialBanner
                styles={{
                  banner: { ...styles.banner, backgroundColor: `${colors.secondary}4A` },
                  bannerLabel: [styles.bannerLabel, typography.label, { color: colors.secondary }],
                }}
              />
              <Controls controlDatas={controlDatas} controlStyles={controlStyles} />
            </View>
          </LinearGradient>
        </ImageBackground>
      </View>
    );
  }

  return null;
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  gradient: {
    width: '100%',
    height: '100%',
  },
  imgBg: {
    width: '100%',
  },
  bottomContainer: {
    alignItems: 'center',
    alignSelf: 'center',
    position: 'absolute',
    bottom: 0,
    padding: spacing.s,
  },
  banner: {
    marginVertical: spacing.s,
    borderRadius: 5,
    paddingVertical: 4,
    paddingHorizontal: 60,
  },
  bannerLabel: {
    textTransform: 'uppercase',
    letterSpacing: 0.5,
  },
});

const controlStyles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    width: '100%',
    justifyContent: 'space-around',
    paddingHorizontal: spacing.xl,
    marginVertical: spacing.xs,
  },
  controlWrapper: {
    padding: spacing.xs,
    alignItems: 'center',
    justifyContent: 'center',
    flex: 1,
  },
});
