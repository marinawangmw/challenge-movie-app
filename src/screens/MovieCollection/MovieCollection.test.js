import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';
import { withProviders } from '@/test-utils';
import { MovieCollection } from '@/screens';
import { movieList, navigation } from '@/test-utils/helper';
import { MovieItem } from '@/components';

const emptyRoute = { params: { collection: [{ movieList: [], title: 'test empty' }] } };
const routeWithData = {
  params: {
    collection: [],
  },
};

describe('Movie Collection renders correcly', () => {
  it('with empty route', () => {
    const tree = renderer
      .create(<MovieCollection route={emptyRoute} navigation={navigation} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('with route with data of movie list', () => {
    const tree = renderer
      .create(<MovieCollection route={routeWithData} navigation={navigation} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('with all items in movie list', () => {
    const component = shallow(<MovieCollection route={routeWithData} navigation={navigation} />);
    expect(component.length).toBe(1);
  });

  it('with no object message', () => {
    const noObjectMessage = 'Nothing found';
    const { queryByText } = render(
      withProviders(
        <MovieCollection
          route={emptyRoute}
          navigation={navigation}
          noObjectMessage={noObjectMessage}
        />
      )
    );

    expect(queryByText('Nothing found')).not.toBeNull();
  });

  it('should render all items', () => {
    const wrapper = shallow(
      withProviders(<MovieCollection route={routeWithData} navigation={navigation} />)
    );

    wrapper.find({ testID: 'movieItem' }).forEach((node, index) => {
      expect(node.props().children).toBe(routeWithData.params.collection[index].title);
    });
  });
});

describe('Movie item component', () => {
  it('renders correctly and match snapshot', () => {
    const item = { item: movieList.movieList[0] };
    const tree = renderer
      .create(<MovieItem item={item} route={emptyRoute} navigation={navigation} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
