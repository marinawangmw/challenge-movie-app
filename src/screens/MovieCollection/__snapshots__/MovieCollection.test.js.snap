// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Collection renders correcly with empty route 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexGrow": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingVertical": 40,
      }
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexGrow": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(28, 28, 30)",
            }
          }
        >
          Nothing here yet
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Movie Collection renders correcly with route with data of movie list 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexGrow": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingVertical": 40,
      }
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexGrow": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(28, 28, 30)",
            }
          }
        >
          Nothing here yet
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Movie item component renders correctly and match snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "height": 150,
        "margin": 1,
        "width": 90,
      }
    }
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "uri": "undefined/aJbeDPUwDe9MLpseTFrNTBaQPzf.jpg",
        }
      }
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": 150,
            "width": 90,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Object {
          "backgroundColor": undefined,
        }
      }
    >
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "rgb(28, 28, 30)",
            },
          ]
        }
      >
        Movy Original
      </Text>
    </View>
  </View>
</View>
`;
